File: perl5/generate-elevenlabs.pl
#!/usr/bin/perl
# generate-elevenlabs.mjs — portiert nach perl5
# Quelle: javascript, Onboarding@main:scripts/generate-elevenlabs.mjs
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use LWP::UserAgent;
use HTTP::Request::Common qw(POST);
use JSON qw(encode_json decode_json);
use File::Path qw(make_path);
use File::Spec;
use Cwd qw(abs_path);

# Umgebungsvariablen abrufen
my $key = $ENV{ELEVENLABS_API_KEY};
my $voiceId = $ENV{ELEVENLABS_VOICE_ID} // "JBFqnCBsd6RMkjVDRZzb";

# Fehler werfen, wenn API-Key fehlt
die "ELEVENLABS_API_KEY fehlt." unless $key;

my $text = "Neun Projekte. Zwei Plattformen. Ein Ort, an dem Ideen verbunden und weiterentwickelt werden.";

# URL für ElevenLabs API zusammenstellen
my $url = "https://api.elevenlabs.io/v1/text-to-speech/$voiceId?output_format=mp3_44100_128";

# User-Agent für HTTP-Anfragen erstellen
my $ua = LWP::UserAgent->new;

# POST-Request vorbereiten
my $req = POST $url,
    Content_Type => 'application/json',
    'xi-api-key' => $key,
    Content => encode_json({
        text => $text,
        model_id => "eleven_multilingual_v2",
        voice_settings => {
            stability => 0.58,
            similarity_boost => 0.72,
            style => 0.18,
            use_speaker_boost => \1  # true in JSON
        }
    });

# Request senden
my $response = $ua->request($req);

# Fehler werfen, wenn Response nicht erfolgreich
die "ElevenLabs fehlgeschlagen: " . $response->code unless $response->is_success;

# Zielverzeichnis bestimmen und ggf. erstellen
my $script_dir = dirname(abs_path(__FILE__));
my $project_root = dirname($script_dir);
my $audio_dir = File::Spec->catdir($project_root, 'public', 'audio');
make_path($audio_dir) unless -d $audio_dir;

# Audiodatei speichern
my $audio_file = File::Spec->catfile($audio_dir, 'project-narration.mp3');
open my $fh, '>', $audio_file or die "Konnte Datei nicht öffnen: $!";
print $fh $response->content;
close $fh;

# JSON-Ergebnisdatei schreiben
my $result_data = {
    model => "eleven_multilingual_v2",
    voiceId => $voiceId,
    characters => length($text),
    text => $text,
    output => "public/audio/project-narration.mp3"
};

my $json_output = JSON->new->pretty->encode($result_data);
my $json_file = File::Spec->catfile($project_root, 'media-production', 'elevenlabs-result.json');
open my $json_fh, '>', $json_file or die "Konnte JSON-Datei nicht öffnen: $!";
print $json_fh $json_output;
close $json_fh;

print "ElevenLabs abgeschlossen: " . length($text) . " Zeichen.\n";

# Hilfsfunktion für dirname (ähnlich node:path)
sub dirname {
    my ($path) = @_;
    $path =~ s/[\/\\][^\/\\]*$//;
    return $path || '.';
}

File: python/generate-elevenlabs.py
#!/usr/bin/env python3
# generate-elevenlabs.mjs — portiert nach python
# Quelle: javascript, Onboarding@main:scripts/generate-elevenlabs.mjs
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

import os
import json
import requests
from pathlib import Path

key = os.environ.get("ELEVENLABS_API_KEY")
voiceId = os.environ.get("ELEVENLABS_VOICE_ID") or "JBFqnCBsd6RMkjVDRZzb"
if not key:
    raise Exception("ELEVENLABS_API_KEY fehlt.")

text = "Neun Projekte. Zwei Plattformen. Ein Ort, an dem Ideen verbunden und weiterentwickelt werden."
response = requests.post(
    f"https://api.elevenlabs.io/v1/text-to-speech/{voiceId}?output_format=mp3_44100_128",
    headers={"xi-api-key": key, "Content-Type": "application/json"},
    json={
        "text": text,
        "model_id": "eleven_multilingual_v2",
        "voice_settings": {
            "stability": 0.58,
            "similarity_boost": 0.72,
            "style": 0.18,
            "use_speaker_boost": True
        }
    }
)

if not response.ok:
    raise Exception(f"ElevenLabs fehlgeschlagen: {response.status_code}")

# Create directories and write files
audio_dir = Path(__file__).parent.parent / "public" / "audio"
audio_dir.mkdir(parents=True, exist_ok=True)

audio_file = audio_dir / "project-narration.mp3"
with open(audio_file, "wb") as f:
    f.write(response.content)

result_file = Path(__file__).parent.parent / "media-production" / "elevenlabs-result.json"
result_data = {
    "model": "eleven_multilingual_v2",
    "voiceId": voiceId,
    "characters": len(text),
    "text": text,
    "output": "public/audio/project-narration.mp3"
}

with open(result_file, "w", encoding="utf-8") as f:
    json.dump(result_data, f, indent=2, ensure_ascii=False)

print(f"ElevenLabs abgeschlossen: {len(text)} Zeichen.")
